From 7abdec724c7b77e6c0508fc47c7bf7a19dd94ace Mon Sep 17 00:00:00 2001
From: Joe Connan <joe.connan@imd-tec.com>
Date: Wed, 17 Jun 2026 14:50:44 +0100
Subject: [PATCH] ath11k: Fix ATH11K-related crash on shutdown (#70)

Performing a reboot or shutdown from userspace or adb, or attempting to
rmmod ath11k_pci on a board connected to a Qualcomm WiFi card that uses
the ATH11K WiFi driver causes the board to crash and produce the
following log:
```
Rebooting.
[   33.472401][    T1] TRACE ATH11k: ath11k_pci_ce_irq_disable
[   33.483382][    T1] TRACE ATH11k: ath11k_pci_ce_irq_disable
[   33.489136][    T1] TRACE ATH11k: ath11k_pci_ce_irq_disable
[   33.494910][    T1] TRACE ATH11k: ath11k_pci_ce_irq_disable
[   33.500658][    T1] TRACE ATH11k: ath11k_pci_ce_irq_disable
[   33.506396][    T1] TRACE ATH11k: ath11k_pci_ce_irq_disable
[   33.512135][    T1] TRACE ATH11k: __ath11k_pci_ext_irq_disable
[   33.518404][    T1] TRACE ATH11k: ath11k_pci_power_down
[   33.562868][    T1] TRACE ATH11k: ath11k_pci_sw_reset pwr_on 0
[   33.568916][    T1] TRACE ATH11k: ath11k_pci_soc_global_reset ab: ffffff8040b00000
[   33.576767][    T1] TRACE ATH11k: ath11k_pci_soc_global_reset read done
[   33.583580][    T1] TRACE ATH11k: ath11k_pci_soc_global_reset write done
[   33.590024][  T493] arm-smmu 15000000.apps-smmu: Unhandled arm-smmu context fault from 0000:01:00.0!
[   33.599867][  T493] arm-smmu 15000000.apps-smmu: FAR    = 0x0000000000000000
[   33.600509][    T1] TRACE ATH11k: ath11k_pci_soc_global_reset write 2 done
[   33.607131][  T493] arm-smmu 15000000.apps-smmu: PAR    = 0x0000000000000000
[   33.607142][  T493] arm-smmu 15000000.apps-smmu: FSR    = 0x40000402 [TF W SS ]
[   33.607153][  T493] arm-smmu 15000000.apps-smmu: FSYNR0    = 0x7e0011
[   33.635629][  T493] arm-smmu 15000000.apps-smmu: FSYNR1    = 0xb300
[   33.642124][  T493] arm-smmu 15000000.apps-smmu: context bank#    = 0x3
[   33.648941][  T493] arm-smmu 15000000.apps-smmu: TTBR0  = 0x0000000000000000
[   33.656238][  T493] arm-smmu 15000000.apps-smmu: TTBR1  = 0x0000000000000000
[   33.663502][  T493] arm-smmu 15000000.apps-smmu: SCTLR  = 0x00c000e7 ACTLR  = 0x00000003
[   33.671875][  T493] arm-smmu 15000000.apps-smmu: CBAR  = 0x0001f300
[   33.678337][  T493] arm-smmu 15000000.apps-smmu: MAIR0   = 0xf404ff44 MAIR1   = 0x0000efe4
[   33.686888][  T493] arm-smmu 15000000.apps-smmu: SID = 0x1401
[   33.692822][  T493] arm-smmu 15000000.apps-smmu: Client info: BID=0x5, PID=0x13, MID=0x0
[   33.701204][  T493] arm-smmu 15000000.apps-smmu: soft iova-to-phys=0x0000000000000000
[   33.709432][  T493] qsmmuv500-tbu 16cd000.pcie_qtb: TBU opted-out for halting!
[   33.717341][  T493] qsmmuv500-tbu 16cd000.pcie_qtb: TBU opted-out for halting!
[   33.725248][  T493] qsmmuv500-tbu 16cd000.pcie_qtb: TBU opted-out for halting!

```
This is caused by the ATH11K driver not being unloaded correctly and
appearing to perform operations on NULL pointers.

This issue can be resolved with a combination of two patches:
[The first patch from the openwrt kernel
driver](https://patchwork.kernel.org/project/linux-wireless/patch/20250416100939.282481-1-balsam.chihi@moment.tech/)
removes issue caused by debug registers being cleared after unloading.

[This second patch from the linux
kernel](https://github.com/torvalds/linux/commit/96527d527b271d950367ad13e3de8b0673545622)
makes sure to disable MSI as well as clearing the MSI address as part of
the shutdown and remove procedures.

---------

Signed-off-by: Baochen Qiang <bqiang@codeaurora.org>
Signed-off-by: Jouni Malinen <jouni@codeaurora.org>
Signed-off-by: Kalle Valo <kvalo@codeaurora.org>
Co-authored-by: host <host@imd-tec.com>
Co-authored-by: Baochen Qiang <bqiang@codeaurora.org>
---
 drivers/net/wireless/ath/ath11k/pci.c | 75 ++++++++++++++++++++-------
 1 file changed, 55 insertions(+), 20 deletions(-)

diff --git a/drivers/net/wireless/ath/ath11k/pci.c b/drivers/net/wireless/ath/ath11k/pci.c
index 768389c1c..ae8fb218d 100644
--- a/drivers/net/wireless/ath/ath11k/pci.c
+++ b/drivers/net/wireless/ath/ath11k/pci.c
@@ -280,16 +280,18 @@ static void ath11k_pci_soc_global_reset(struct ath11k_base *ab)
 		ath11k_warn(ab, "link down error during global reset\n");
 }
 
-static void ath11k_pci_clear_dbg_registers(struct ath11k_base *ab)
+static void ath11k_pci_clear_dbg_registers(struct ath11k_base *ab, bool power_on)
 {
 	u32 val;
 
-	/* read cookie */
-	val = ath11k_pci_read32(ab, PCIE_Q6_COOKIE_ADDR);
-	ath11k_dbg(ab, ATH11K_DBG_PCI, "cookie:0x%x\n", val);
+	if (power_on) {
+		/* read cookie */
+		val = ath11k_pci_read32(ab, PCIE_Q6_COOKIE_ADDR);
+		ath11k_dbg(ab, ATH11K_DBG_PCI, "cookie:0x%x\n", val);
 
-	val = ath11k_pci_read32(ab, WLAON_WARM_SW_ENTRY);
-	ath11k_dbg(ab, ATH11K_DBG_PCI, "WLAON_WARM_SW_ENTRY 0x%x\n", val);
+		val = ath11k_pci_read32(ab, WLAON_WARM_SW_ENTRY);
+		ath11k_dbg(ab, ATH11K_DBG_PCI, "WLAON_WARM_SW_ENTRY 0x%x\n", val);
+	}
 
 	/* TODO: exact time to sleep is uncertain */
 	mdelay(10);
@@ -300,14 +302,16 @@ static void ath11k_pci_clear_dbg_registers(struct ath11k_base *ab)
 	ath11k_pci_write32(ab, WLAON_WARM_SW_ENTRY, 0);
 	mdelay(10);
 
-	val = ath11k_pci_read32(ab, WLAON_WARM_SW_ENTRY);
-	ath11k_dbg(ab, ATH11K_DBG_PCI, "WLAON_WARM_SW_ENTRY 0x%x\n", val);
-
-	/* A read clear register. clear the register to prevent
-	 * Q6 from entering wrong code path.
-	 */
-	val = ath11k_pci_read32(ab, WLAON_SOC_RESET_CAUSE_REG);
-	ath11k_dbg(ab, ATH11K_DBG_PCI, "soc reset cause:%d\n", val);
+	if (power_on) {
+		val = ath11k_pci_read32(ab, WLAON_WARM_SW_ENTRY);
+		ath11k_dbg(ab, ATH11K_DBG_PCI, "WLAON_WARM_SW_ENTRY 0x%x\n", val);
+
+		/* A read clear register. clear the register to prevent
+		 * Q6 from entering wrong code path.
+		 */
+		val = ath11k_pci_read32(ab, WLAON_SOC_RESET_CAUSE_REG);
+		ath11k_dbg(ab, ATH11K_DBG_PCI, "soc reset cause:%d\n", val);
+	}
 }
 
 static int ath11k_pci_set_link_reg(struct ath11k_base *ab,
@@ -445,7 +449,7 @@ static void ath11k_pci_sw_reset(struct ath11k_base *ab, bool power_on)
 	ath11k_mhi_clear_vector(ab);
 	ath11k_pci_soc_global_reset(ab);
 	ath11k_mhi_set_mhictrl_reset(ab);
-	ath11k_pci_clear_dbg_registers(ab);
+	ath11k_pci_clear_dbg_registers(ab, power_on);
 }
 
 int ath11k_pci_get_msi_irq(struct device *dev, unsigned int vector)
@@ -863,7 +867,32 @@ static void ath11k_pci_ce_irqs_enable(struct ath11k_base *ab)
 	}
 }
 
-static int ath11k_pci_enable_msi(struct ath11k_pci *ab_pci)
+static void ath11k_pci_msi_config(struct ath11k_pci *ab_pci, bool enable)
+{
+	struct pci_dev *dev = ab_pci->pdev;
+	u16 control;
+
+	pci_read_config_word(dev, dev->msi_cap + PCI_MSI_FLAGS, &control);
+
+	if (enable)
+		control |= PCI_MSI_FLAGS_ENABLE;
+	else
+		control &= ~PCI_MSI_FLAGS_ENABLE;
+
+	pci_write_config_word(dev, dev->msi_cap + PCI_MSI_FLAGS, control);
+}
+
+static void ath11k_pci_msi_enable(struct ath11k_pci *ab_pci)
+{
+	ath11k_pci_msi_config(ab_pci, true);
+}
+
+static void ath11k_pci_msi_disable(struct ath11k_pci *ab_pci)
+{
+	ath11k_pci_msi_config(ab_pci, false);
+}
+
+static int ath11k_pci_alloc_msi(struct ath11k_pci *ab_pci)
 {
 	struct ath11k_base *ab = ab_pci->ab;
 	const struct ath11k_msi_config *msi_config = ab_pci->msi_config;
@@ -884,6 +913,7 @@ static int ath11k_pci_enable_msi(struct ath11k_pci *ab_pci)
 		else
 			return num_vectors;
 	}
+	ath11k_pci_msi_disable(ab_pci);
 
 	msi_desc = irq_get_msi_desc(ab_pci->pdev->irq);
 	if (!msi_desc) {
@@ -906,7 +936,7 @@ static int ath11k_pci_enable_msi(struct ath11k_pci *ab_pci)
 	return ret;
 }
 
-static void ath11k_pci_disable_msi(struct ath11k_pci *ab_pci)
+static void ath11k_pci_free_msi(struct ath11k_pci *ab_pci)
 {
 	pci_free_irq_vectors(ab_pci->pdev);
 }
@@ -1029,6 +1059,8 @@ static int ath11k_pci_power_up(struct ath11k_base *ab)
 	 */
 	ath11k_pci_aspm_disable(ab_pci);
 
+	ath11k_pci_msi_enable(ab_pci);
+
 	ret = ath11k_mhi_start(ab_pci);
 	if (ret) {
 		ath11k_err(ab, "failed to start mhi: %d\n", ret);
@@ -1049,6 +1081,9 @@ static void ath11k_pci_power_down(struct ath11k_base *ab)
 	ath11k_pci_aspm_restore(ab_pci);
 
 	ath11k_pci_force_wake(ab_pci->ab);
+
+	ath11k_pci_msi_disable(ab_pci);
+
 	ath11k_mhi_stop(ab_pci);
 	clear_bit(ATH11K_PCI_FLAG_INIT_DONE, &ab_pci->flags);
 	ath11k_pci_sw_reset(ab_pci->ab, false);
@@ -1293,7 +1328,7 @@ static int ath11k_pci_probe(struct pci_dev *pdev,
 		goto err_pci_free_region;
 	}
 
-	ret = ath11k_pci_enable_msi(ab_pci);
+	ret = ath11k_pci_alloc_msi(ab_pci);
 	if (ret) {
 		ath11k_err(ab, "failed to enable msi: %d\n", ret);
 		goto err_pci_free_region;
@@ -1347,7 +1382,7 @@ static int ath11k_pci_probe(struct pci_dev *pdev,
 	ath11k_mhi_unregister(ab_pci);
 
 err_pci_disable_msi:
-	ath11k_pci_disable_msi(ab_pci);
+	ath11k_pci_free_msi(ab_pci);
 
 err_pci_free_region:
 	ath11k_pci_free_region(ab_pci);
@@ -1378,7 +1413,7 @@ static void ath11k_pci_remove(struct pci_dev *pdev)
 	ath11k_mhi_unregister(ab_pci);
 
 	ath11k_pci_free_irq(ab);
-	ath11k_pci_disable_msi(ab_pci);
+	ath11k_pci_free_msi(ab_pci);
 	ath11k_pci_free_region(ab_pci);
 
 	ath11k_hal_srng_deinit(ab);
-- 
2.41.0

